src/tests.rs

```diff
@@ -0,0 +1,136 @@
+use crate::twin::TwinDB;
+
+async fn start_rmb(
+    db: SubstrateTwinDB<MemCache<Twin>>,
+    storage: RedisStorage,
+    ident: identity::Ed25519Signer,
+    address: String,
+) -> Result<()> {
+    let processor_handler_1 = tokio::spawn(processor(1, storage.clone()));
+
+    let api_handler_1 =
+        tokio::spawn(HttpApi::new(1, address, storage.clone(), ident.clone(), db.clone())?.run());
+
+    let proxy_handler_1 = tokio::spawn(ProxyWorker::new(1, 10, storage.clone(), db.clone()).run());
+
+    let workers_handler_1 = tokio::task::spawn(HttpWorker::new(1000, storage, db, ident).run());
+
+    tokio::select! {
+        result = processor_handler_1 => {
+            if let Err(err) = result {
+                bail!("message processor panicked unexpectedly: {}", err);
+            }
+        },
+        result = api_handler_1 => {
+            match result {
+                Err(err) => bail!("http server panicked unexpectedly: {}", err),
+                Ok(Ok(_)) => bail!("http server exited unexpectedly"),
+                Ok(Err(err)) => bail!("http server exited with error: {}", err),
+            }
+        },
+        result = workers_handler_1 => {
+            if let Err(err) = result {
+                bail!("http workers panicked unexpectedly: {}", err);
+            }
+        },
+        result = proxy_handler_1 => {
+            if let Err(err) = result {
+                bail!("proxy workers panicked unexpectedly: {}", err);
+            }
+        },
+    };
+    unreachable!();
+}
+
+#[tokio::test]
+async fn test_end_to_end() {
+    const WORDS1: &str = "neck stage box cup core magic produce exercise happy rely vocal then";
+    const WORDS2: &str =
+        "nominee slow sting tell point bleak sheriff outer push visual basket grief";
+    // load from WORD
+    let t1 = identity::Ed25519Signer::try_from(WORDS1).unwrap();
+    let t2 = identity::Ed25519Signer::try_from(WORDS2).unwrap();
+    // get account id
+    let acc1 = identity::Signers::Ed25519(t1.clone()).account();
+    let acc2 = identity::Signers::Ed25519(t2.clone()).account();
+    // create cache
+    let mem: MemCache<Twin> = MemCache::new();
+    // create dummy entities for testing
+    let twin1: Twin = Twin {
+        version: 1,
+        id: 1,
+        account: acc1.clone(),
+        address: "127.0.0.1:5810".to_string(),
+        entities: vec![],
+    };
+    let twin2: Twin = Twin {
+        version: 1,
+        id: 2,
+        account: acc2.clone(),
+        address: "127.0.0.1:5820".to_string(),
+        entities: vec![],
+    };
+    // insert dummy entities into cache
+    mem.set(1, twin1.clone())
+        .await
+        .context("can not set value to cache")
+        .unwrap();
+    mem.set(2, twin2.clone())
+        .await
+        .context("can not set value to cache")
+        .unwrap();
+
+    // create db
+    let db1 =
+        SubstrateTwinDB::<MemCache<Twin>>::new("wss://tfchain.dev.grid.tf", mem.clone()).unwrap();
```

Comment:
Why are u abusing SubstrateTwinDB. why not implement InMemoryDB like i showed you?

Reply:
because I think it is easier to use what is already implemented than reimplementing something similar (that will give the same expected result).

Reply:
Okay, Will do.